File: javascript/benchmark.js
const crypto = require('crypto');

function runAllBenchmarks() {
    const results = [];
    
    console.log("Running CPU-intensive benchmarks...");
    results.push(benchmarkPrimeNumbers());
    results.push(benchmarkMatrixMultiplication());
    results.push(benchmarkCryptographicHashing());
    results.push(benchmarkMathOperations());
    
    console.log("Running memory-intensive benchmarks...");
    results.push(benchmarkLargeArraySort());
    results.push(benchmarkMemoryAllocation());
    results.push(benchmarkStringConcatenation());
    
    return results;
}

function benchmarkPrimeNumbers() {
    const start = process.hrtime.bigint();
    
    let count = 0;
    const limit = 100000;
    for (let i = 2; i <= limit; i++) {
        if (isPrime(i)) {
            count++;
        }
    }
    
    const duration = process.hrtime.bigint() - start;
    const durationSeconds = Number(duration) / 1e9;
    
    return {
        test: 'Prime Numbers (up to 100k)',
        duration_ns: Number(duration),
        memory_bytes: 0,  // JavaScriptでは正確なメモリ測定が困難
        operations: count,
        ops_per_sec: count / durationSeconds
    };
}

function isPrime(n) {
    if (n < 2) return false;
    for (let i = 2; i <= Math.sqrt(n); i++) {
        if (n % i === 0) return false;
    }
    return true;
}

function benchmarkMatrixMultiplication() {
    const start = process.hrtime.bigint();
    
    const size = 500;
    const a = Array(size).fill().map(() => Array(size).fill().map(() => Math.random()));
    const b = Array(size).fill().map(() => Array(size).fill().map(() => Math.random()));
    const c = Array(size).fill().map(() => Array(size).fill(0));
    
    for (let i = 0; i < size; i++) {
        for (let j = 0; j < size; j++) {
            for (let k = 0; k < size; k++) {
                c[i][j] += a[i][k] * b[k][j];
            }
        }
    }
    
    const duration = process.hrtime.bigint() - start;
    const durationSeconds = Number(duration) / 1e9;
    const operations = size * size * size;
    
    return {
        test: 'Matrix Multiplication (500x500)',
        duration_ns: Number(duration),
        memory_bytes: 0,
        operations,
        ops_per_sec: operations / durationSeconds
    };
}

function benchmarkCryptographicHashing() {
    const start = process.hrtime.bigint();
    
    const data = Buffer.alloc(1024);
    for (let i = 0; i < 1024; i++) {
        data[i] = Math.floor(Math.random() * 256);
    }
    
    const iterations = 50000;
    for (let i = 0; i < iterations; i++) {
        crypto.createHash('sha256').update(data).digest('hex');
    }
    
    const duration = process.hrtime.bigint() - start;
    const durationSeconds = Number(duration) / 1e9;
    
    return {
        test: 'SHA256 Hashing (50k iterations)',
        duration_ns: Number(duration),
        memory_bytes: 0,
        operations: iterations,
        ops_per_sec: iterations / durationSeconds
    };
}

function benchmarkMathOperations() {
    const start = process.hrtime.bigint();
    
    const iterations = 10000000;
    let result = 0.0;
    for (let i = 0; i < iterations; i++) {
        const x = i;
        result += Math.sin(x) * Math.cos(x) * Math.sqrt(x + 1);
    }
    
    const duration = process.hrtime.bigint() - start;
    const durationSeconds = Number(duration) / 1e9;
    
    // resultを使用して最適化を防ぐ
    if (isNaN(result)) {
        console.log("Unexpected NaN result");
    }
    
    return {
        test: 'Math Operations (10M iterations)',
        duration_ns: Number(duration),
        memory_bytes: 0,
        operations: iterations,
        ops_per_sec: iterations / durationSeconds
    };
}

function benchmarkLargeArraySort() {
    const start = process.hrtime.bigint();
    
    const size = 1000000;
    const data = Array(size).fill().map(() => Math.floor(Math.random() * 1000001));
    
    data.sort((a, b) => a - b);
    
    const duration = process.hrtime.bigint() - start;
    const durationSeconds = Number(duration) / 1e9;
    
    return {
        test: 'Large Array Sort (1M elements)',
        duration_ns: Number(duration),
        memory_bytes: 0,
        operations: size,
        ops_per_sec: size / durationSeconds
    };
}

function benchmarkMemoryAllocation() {
    const start = process.hrtime.bigint();
    
    const allocations = 100000;
    const slices = [];
    
    for (let i = 0; i < allocations; i++) {
        const sliceData = Array(256).fill(i % 256); // 1KB相当のデータ
        slices.push(sliceData);
    }
    
    const duration = process.hrtime.bigint() - start;
    const durationSeconds = Number(duration) / 1e9;
    
    return {
        test: 'Memory Allocation (100k x 1KB)',
        duration_ns: Number(duration),
        memory_bytes: 0,
        operations: allocations,
        ops_per_sec: allocations / durationSeconds
    };
}

function benchmarkStringConcatenation() {
    const start = process.hrtime.bigint();
    
    const iterations = 50000;
    let result = "";
    for (let i = 0; i < iterations; i++) {
        result += `iteration_${i}_`;
    }
    
    const duration = process.hrtime.bigint() - start;
    const durationSeconds = Number(duration) / 1e9;
    
    return {
        test: 'String Concatenation (50k iterations)',
        duration_ns: Number(duration),
        memory_bytes: 0,
        operations: iterations,
        ops_per_sec: iterations / durationSeconds
    };
}

module.exports = {
    runAllBenchmarks
};

File: python3/benchmark.py
#!/usr/bin/env python3

import time
import random
import hashlib
import math

def run_all_benchmarks():
    results = []
    
    print("Running CPU-intensive benchmarks...")
    results.append(benchmark_prime_numbers())
    results.append(benchmark_matrix_multiplication())
    results.append(benchmark_cryptographic_hashing())
    results.append(benchmark_math_operations())
    
    print("Running memory-intensive benchmarks...")
    results.append(benchmark_large_array_sort())
    results.append(benchmark_memory_allocation())
    results.append(benchmark_string_concatenation())
    
    return results

def benchmark_prime_numbers():
    start = time.time()
    
    count = 0
    limit = 100000
    for i in range(2, limit + 1):
        if is_prime(i):
            count += 1
    
    duration = time.time() - start
    
    return {
        'test': 'Prime Numbers (up to 100k)',
        'duration_ns': int(duration * 1e9),
        'memory_bytes': 0,  # Python3では正確なメモリ測定が困難
        'operations': count,
        'ops_per_sec': count / duration
    }

def is_prime(n):
    if n < 2:
        return False
    for i in range(2, int(math.sqrt(n)) + 1):
        if n % i == 0:
            return False
    return True

def benchmark_matrix_multiplication():
    start = time.time()
    
    size = 500
    a = [[random.random() for _ in range(size)] for _ in range(size)]
    b = [[random.random() for _ in range(size)] for _ in range(size)]
    c = [[0.0 for _ in range(size)] for _ in range(size)]
    
    for i in range(size):
        for j in range(size):
            for k in range(size):
                c[i][j] += a[i][k] * b[k][j]
    
    duration = time.time() - start
    operations = size * size * size
    
    return {
        'test': 'Matrix Multiplication (500x500)',
        'duration_ns': int(duration * 1e9),
        'memory_bytes': 0,
        'operations': operations,
        'ops_per_sec': operations / duration
    }

def benchmark_cryptographic_hashing():
    start = time.time()
    
    data = bytes([random.randint(0, 255) for _ in range(1024)])
    
    iterations = 50000
    for _ in range(iterations):
        hashlib.sha256(data).hexdigest()
    
    duration = time.time() - start
    
    return {
        'test': 'SHA256 Hashing (50k iterations)',
        'duration_ns': int(duration * 1e9),
        'memory_bytes': 0,
        'operations': iterations,
        'ops_per_sec': iterations / duration
    }

def benchmark_math_operations():
    start = time.time()
    
    iterations = 10000000
    result = 0.0
    for i in range(iterations):
        x = float(i)
        result += math.sin(x) * math.cos(x) * math.sqrt(x + 1)
    
    duration = time.time() - start
    
    return {
        'test': 'Math Operations (10M iterations)',
        'duration_ns': int(duration * 1e9),
        'memory_bytes': 0,
        'operations': iterations,
        'ops_per_sec': iterations / duration
    }

def benchmark_large_array_sort():
    start = time.time()
    
    size = 1000000
    data = [random.randint(0, 1000000) for _ in range(size)]
    data.sort()
    
    duration = time.time() - start
    
    return {
        'test': 'Large Array Sort (1M elements)',
        'duration_ns': int(duration * 1e9),
        'memory_bytes': 0,
        'operations': size,
        'ops_per_sec': size / duration
    }

def benchmark_memory_allocation():
    start = time.time()
    
    allocations = 100000
    slices = []
    
    for i in range(allocations):
        slice_data = [i % 256] * 256  # 1KB相当のデータ
        slices.append(slice_data)
    
    duration = time.time() - start
    
    return {
        'test': 'Memory Allocation (100k x 1KB)',
        'duration_ns': int(duration * 1e9),
        'memory_bytes': 0,
        'operations': allocations,
        'ops_per_sec': allocations / duration
    }

def benchmark_string_concatenation():
    start = time.time()
    
    iterations = 50000
    result = ""
    for i in range(iterations):
        result += f"iteration_{i}_"
    
    duration = time.time() - start
    
    return {
        'test': 'String Concatenation (50k iterations)',
        'duration_ns': int(duration * 1e9),
        'memory_bytes': 0,
        'operations': iterations,
        'ops_per_sec': iterations / duration
    }